themes/default/content/blog/pulumi-watch-mode-with-rust/index.md

```diff
@@ -0,0 +1,150 @@
+with new APIs. In each of these cases, engineers find value in being able to focus on writing code,
+not typing console commands.
+
+That's why [watch mode exists]({{< relref "/blog/pulumi-watch-mode-fast-inner-loop-development-for-cloud-infrastructure" >}})
+and is one of my favorite features, and it's why I was disappointed to learn this feature [was not
+available to users on Apple Silicon Macs](https://github.com/pulumi/pulumi/pull/6497). As a Windows
+and Linux user, I discovered this while root causing a failing build on one of our providers to the
+library used to implement watch mode. Thus began a deep dive into cgo, file-watching libraries in
+Go, operating system event APIs, and finding a solution for our build issues and missing feature.
+
+To close this feature gap for one of our most important platforms, Pulumi v3.39.0 includes a binary
+built with Rust to implement file-watching. From a user perspective, “pulumi watch” works as it
+always has. And as the first component to use the Rust language, it validated Pulumi's approach to
+engineering of using the best languages and tools available.
+
+## But first, what is and why use watch mode?
+
+Watch mode monitors a project folder and automatically runs an equivalent to `pulumi up` with the
+appropriate flags to skip prompts and displays simplified output. To see how quickly this can work,
+I wrote a Pulumi TypeScript program to deploy an S3 bucket and bucket object "test.txt" before
+writing a timestamp to the same file. That allows me to measure the inner loop with `pulumi watch`
+or using an external program to run “pulumi up”. On each deployment, the program changes one of the
+files in the directory kicking off the next deployment:
+
+```typescript
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+import * as fs from 'fs/promises'
+
+const bucket = new aws.s3.Bucket("my-bucket");
+
+const bucketObject = new aws.s3.BucketObject("my-bucket-object", {
+    bucket: bucket.id,
+    key: "test.txt",
+    source: new pulumi.asset.FileAsset("test.txt"),
+});
+
+bucketObject.content.apply(() => {
+  fs.appendFile('test.txt', `${new Date().valueOf()}\n`);
+});
+```
+
+On a laptop with this simple program, `pulumi watch` consistently completed up to 50% faster, at 2.1
+seconds per iteration than when using an external tool. Watch mode was more consistent, too: the
+standard deviation over 50 samples was reduced to 46ms from 216ms. It does this while being easier
+to remember and therefore use correctly, and with more terse output tuned for the use case of
+quickly iterating on changes to a stack.
+
+Watch mode has changed how I write infrastructure as code, and it brings that benefit to everyone
+without needing to install and learn third-party tools for file-watching.
+
+## The sad state of Go file-watching
+
+Returning to the cross-compilation issue, the team found a provider's builds failed after updating
```

Comment:
provider's builds is kinda difficult to read, prolly a nit.

Reply:
Yeah, I think I might nix "provider" here & in the earlier description or figure out a way to reword it.

Reply:
The word provider was redundant here with the following sentence, cleaned up.